File: src/main/java/org/interview/twitter/model/TwitterMessage.java
package org.interview.twitter.model;

import javax.persistence.*;
import java.util.Date;
import java.util.Set;

@Entity
@Table(name = "MESSAGE")
public class TwitterMessage implements Comparable{

    @Id
    @Column(name = "MESSAGE_ID")
    private Long messageId;

    @Column(name = "CREATION_DATE")
    private Date creationDate;

    @Column(name = "MESSAGE_TEXT")
    private String messageText;

    @Column(name = "AUTHOR")
    @ManyToMany
    @JoinTable(name = "AUTHOR_MESSAGE",
            joinColumns = @JoinColumn(name = "MESSAGE_ID"),
            inverseJoinColumns = @JoinColumn(name = "AUTHOR_ID"))
    private Set<Author> authors;

    public Long getMessageId() {
        return messageId;
    }

    public Date getCreationDate() {
        return creationDate;
    }

    public String getMessageText() {
        return messageText;
    }



    public int compareTo(Object o) {
        final int EQUAL = 0;
        int result;

        TwitterMessage oTwitterMessage = (TwitterMessage) o;

        if(this == oTwitterMessage)
            result = EQUAL;
        else
            result = this.getCreationDate().compareTo(oTwitterMessage.getCreationDate());

        return result;
    }

    public Set<Author> getAuthors() {
        return authors;
    }

    public void setAuthors(Set<Author> authors) {
        this.authors = authors;
    }

    public void setMessageId(Long messageId) {
        this.messageId = messageId;
    }

    public void setCreationDate(Date creationDate) {
        this.creationDate = creationDate;
    }

    public void setMessageText(String messageText) {
        this.messageText = messageText;
    }
}

File: src/main/kotlin/org/interview/twitter/model/TwitterMessage.kt
package org.interview.twitter.model

import java.util.*
import javax.persistence.*

/**
 * @authors Mehdi Najafian
 */
@Entity
@Table(name = "MESSAGE")
class TwitterMessage : Comparable<TwitterMessage> {
    @Id
    @Column(name = "MESSAGE_ID")
    var messageId: Long? = null
    @Column(name = "CREATION_DATE")
    var creationDate: Date? = null
    @Column(name = "MESSAGE_TEXT")
    var messageText: String? = null
    @OneToOne(cascade = [CascadeType.ALL])
    @JoinColumn(name = "USER_ID", referencedColumnName = "USER_ID", unique = false)
    var author: Author? = null

    //this constructor is for Jpa
    constructor()

    constructor(messageId: Long?, creationDate: Date?, messageText: String?, author: Author?) {
        this.messageId = messageId
        this.creationDate = creationDate
        this.messageText = messageText
        this.author = author
    }

    override fun compareTo(other: TwitterMessage): Int {
        var result = 0;
        if (other !== this)
            return this.creationDate!!.compareTo(other.creationDate)
        return result;
    }
}
